documentation/development-docs/patterns-for-standard-cmdlets.md

```diff
@@ -0,0 +1,996 @@
+# Expected Patterns for Standard Cmdlets
+
+## Table of Contents
+
+- [Top-level Resources](#top---level-resources)
+    - [`Get-*` cmdlet](#get--cmdlet)
+    - [`New-*` cmdlet](#new--cmdlet)
+    - [`Remove-*` cmdlet](#remove--cmdlet)
+    - [`Set-*` cmdlet](#set--cmdlet)
+    - [`Update-*` cmdlet](#update--cmdlet)
+- [Child Resources](#child-resources)
+    - [`Get-*` cmdlet](#get--cmdlet-1)
+    - [`New-*` cmdlet](#new--cmdlet-1)
+    - [`Remove-*` cmdlet](#remove--cmdlet-1)
+    - [`Set-*` cmdlet](#set--cmdlet-1)
+    - [`Update-*` cmdlet](#update--cmdlet-1)
+
+## Top-level Resources
+
+### `Get-*` cmdlet
+
+All top-level resources should have a `Get-*` cmdlet that allows users to list the resources in their subscription or a resource group, as well as get a specific resource. In addition, users should be able to provide the resource id of the resource they want to get, and the cmdlet will parse the string to get the necessary identity information.
+
+#### Parameter sets
+
+To enable the scenarios mentioned previously, the cmdlet will need three parameter sets:
+
+```
+Get-AzureRmFoo [-ResourceGroupName <String>]
+
+Get-AzureRmFoo -ResourceGroupName <String> -Name <String>
+
+Get-AzureRmFoo -ResourceId <String>
+```
+
+The first parameter set has an optional `-ResourceGroupName` parameter, which allows the user to list all resources in a subscription or in a resource group. The second parameter set has required `-ResourceGroupName` and `-Name` parameters, which allows the user to get a specific resource. The third parameter set has a required `-ResourceId` parameter, which allows the user to get a specific resource by resource id.
+
+#### C# example
+
+<details><summary>Click to expand example</summary>
+<p>
+
+```cs
+namespace Microsoft.Azure.Commands.Foo
+{
+    [Cmdlet(VerbsCommon.Get, "AzureRmFoo", DefaultParameterSetName = "ListParameterSet"), OutputType(typeof(PSFoo))]
+    public class GetFooCommand : FooBaseCmdlet
+    {
+        [Parameter(Mandatory = false, ParameterSetName = "ListParameterSet")]
+        [Parameter(Mandatory = true, ParameterSetName = "GetByFooNameParameterSet")]
+        [ResourceGroupCompleter]
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = true, ParameterSetName = "GetByFooNameParameterSet")]
+        [ValidateNotNullOrEmpty]
+        public string Name { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipelineByPropertyName = true, ParameterSetName = "GetByResourceIdParameterSet" )]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        public override void ExecuteCmdlet()
+        {
+            if (this.IsParameterBound(c => c.ResourceId))
```

Comment:
why not use the parameter set names instead to choose which parameter has been used?
(for example [here](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Billing/Commands.Billing/Cmdlets/BillingPeriods/GetAzureRmBillingPeriod.cs#L39))

Reply:
It's a matter of preference -- both accomplish the same thing as long as the parameter you are checking that's bound is the unique mandatory parameter in the parameter set that you'd be checking. I also believe it's better practice to check if a parameter is bound since there's a difference between a parameter being `null` because the user didn't provide a value and the user passing the `null` value itself. Using this method will be able to differentiate the two.